lib/corefoundation/preferences.rb

```diff
@@ -13,55 +10,120 @@ module CF
-  attach_function 'CFPreferencesCopyKeyList', [:application_id, :username, :hostname], :cfarrayref
 
   attach_function 'CFPreferencesSetAppValue', [:key, :value, :application_id], :void
   attach_function 'CFPreferencesSetValue', [:key, :value, :application_id, :username, :hostname], :void
+
   attach_function 'CFPreferencesAppSynchronize', [:application_id], :bool
 
+  # Interface to the preference utilities from Corefoundation.framework.
+  # Documentation at https://developer.apple.com/documentation/corefoundation/preferences_utilities
+  #
   class Preferences
     CURRENT_USER = CF.kCFPreferencesCurrentUser
     ALL_USERS = CF.kCFPreferencesAnyUser
     CURRENT_HOST = CF.kCFPreferencesCurrentHost
     ALL_HOSTS = CF.kCFPreferencesAnyHost
 
     class << self
+      # Returns the output from `CFPreferencesCopyValue` call after converting it to ruby type.
+      #
+      # @param [String] key Preference key to read
+      # @param [String] application_id Preference domain for the key
+      # @param [String, Symbol] username Domain user (current, any or a specific user)
+      # @param [String, Symbol] hostname Hostname (current, all hosts or a specific host)
+      #
+      # @return [VALUE] Preference value returned from the `CFPreferencesCopyValue` call.
+      #
       def get(key, application_id, username = nil, hostname = nil)
-        # TODO: Check usecase for `CFPreferencesCopyAppValue`
-
-        plist_ref = if username && hostname
-                      CF.CFPreferencesCopyValue(key.to_cf, application_id.to_cf, arg_to_cf(username), arg_to_cf(hostname))
-                    else
+        plist_ref = if username.nil? or hostname.nil?
                       CF.CFPreferencesCopyAppValue(key.to_cf, application_id.to_cf)
+                    else
+                      CF.CFPreferencesCopyValue(
+                        key.to_cf,
+                        application_id.to_cf,
+                        arg_to_cf(username),
+                        arg_to_cf(hostname)
+                      )
                     end
         CF::Base.typecast(plist_ref).to_ruby unless plist_ref.null?
       end
 
+      # Calls the {#self.get} method and raise a `PreferenceDoesNotExist` error if `nil` is returned.
+      #
+      # @param [String] key Preference key to read
+      # @param [String] application_id Preference domain for the key
+      # @param [String, Symbol] username Domain user (current, any or a specific user)
+      # @param [String, Symbol] hostname Hostname (current, all hosts or a specific host)
+      #
+      # @raise [PreferenceDoesNotExist] If returned value is nil.
+      #
+      # @return [VALUE] Preference value returned from the {#self.get} method call.
+      #
       def get!(key, application_id, username = nil, hostname = nil)
-        get(key, application_id, username, hostname) || (raise CF::Error, "Preference key `#{key}` not found")
+        value = get(key, application_id, username, hostname)
+        if value.nil?
```

Comment:
```ruby
get(key, application_id, username, hostname) || raise(CF::Exceptions::PreferenceDoesNotExist.new(key, application_id, hostname))
```
was cleaner, we can keep that

Reply:
What you are suggesting will raise an error for booleans too. We would only want to raise it if it's nil.

Reply:
If that function could potentially return boolean, it shouldn't. It should only return either the value or nil

Reply:
domain properties can have boolean values.